File: backends/java/zappy-application/src/main/java/com/zappymart/application/accounts/RevokeSession.java
package com.zappymart.application.accounts;

import com.zappymart.application.Visitor;
import com.zappymart.application.ports.Clock;
import com.zappymart.application.ports.SessionStore;
import com.zappymart.application.ports.UnitOfWork;
import com.zappymart.domain.accounts.Session;
import com.zappymart.domain.shared.Result;
import com.zappymart.domain.shared.UserErrorCode;

import java.time.Instant;
import java.util.List;
import java.util.Optional;

public final class RevokeSession {

    private final UnitOfWork unitOfWork;
    private final SessionStore sessionStore;
    private final Clock clock;

    public RevokeSession(UnitOfWork unitOfWork, SessionStore sessionStore, Clock clock) {
        this.unitOfWork = unitOfWork;
        this.sessionStore = sessionStore;
        this.clock = clock;
    }

    public Result<List<Session>> execute(Visitor visitor, String sessionId) {
        if (!visitor.isSignedIn()) {
            return Result.refuse(UserErrorCode.NOT_AUTHENTICATED,
                    "Sign in to manage your sessions.");
        }
        return unitOfWork.inTransaction(() -> {
            Instant moment = clock.now();
            Optional<Session> session = sessionStore.sessionById(sessionId)
                    .filter(found -> found.customerId().equals(visitor.customerId()))
                    .filter(found -> found.isOpenAt(moment));
            if (session.isEmpty()) {
                return Result.refuse(UserErrorCode.SESSION_NOT_FOUND,
                        "You have no open session with that id.", "sessionId");
            }
            sessionStore.revokeFamily(session.get().id(), moment);
            return Result.of(sessionStore.openSessionsOf(visitor.customerId(), moment));
        });
    }
}

File: backends/kotlin/zappy-application/src/main/kotlin/nl/zappymart/application/accounts/RevokeSession.kt
package nl.zappymart.application.accounts

import nl.zappymart.application.Visitor
import nl.zappymart.application.ports.Clock
import nl.zappymart.application.ports.SessionRepository
import nl.zappymart.application.ports.UnitOfWork
import nl.zappymart.domain.accounts.Session
import nl.zappymart.domain.shared.Result
import nl.zappymart.domain.shared.UserErrorCode
import nl.zappymart.domain.shared.refusal

class RevokeSession(
    private val sessions: SessionRepository,
    private val clock: Clock,
    private val unitOfWork: UnitOfWork,
) {

    fun execute(visitor: Visitor, sessionId: String): Result<List<Session>> {
        val customerId = visitor.customerId
            ?: return refusal(UserErrorCode.NOT_AUTHENTICATED, "Revoking a session needs a signed in customer.")
        return unitOfWork.execute {
            val moment = clock.moment()
            val session = sessions.findById(sessionId)
            if (session == null || session.customerId != customerId) {
                refusal(UserErrorCode.SESSION_NOT_FOUND, "That session does not belong to you.", "sessionId")
            } else {
                sessions.revokeSessionAndItsTokens(sessionId, moment)
                Result.Success(sessions.findOpenForCustomer(customerId, moment))
            }
        }
    }
}
